File: backend/aiHelper.js
import { OpenAI } from 'openai';
import { config } from './config.js';

let openaiClient = null;
if (config.openaiApiKey) {
  openaiClient = new OpenAI({
  apiKey: config.openaiApiKey,
  baseURL: "https://openrouter.ai/api/v1",
});
}

const SCHEMA_PROMPT = `
You are an expert SQL query generator for an Apparel & Textiles ERP system.
The system uses SQLite (locally) or PostgreSQL (in production). Write SQL queries that work on both.
Return ONLY the raw SQL query. Do not wrap it in markdown code blocks, do not use backticks, and do not provide any explanations.

Database Schema:

1. suppliers
- id (INTEGER PRIMARY KEY)
- name (TEXT) -- Company Name
- contact_email (TEXT)
- phone (TEXT)
- address (TEXT)
- country (TEXT)
- contact (TEXT)
- lead_time (INTEGER) -- Lead time in days
- rating (REAL)
- created_at (TIMESTAMP)

2. buyers
- id (INTEGER PRIMARY KEY)
- name (TEXT) -- Company Name
- contact_email (TEXT)
- phone (TEXT)
- address (TEXT)
- country (TEXT)
- buyer_category (TEXT)
- created_at (TIMESTAMP)

3. finished_goods
- style_number (TEXT PRIMARY KEY)
- style_name (TEXT)
- category (TEXT)
- fabric (TEXT)
- gsm (INTEGER)
- color (TEXT)
- print (TEXT)
- season (TEXT)
- brand (TEXT)
- cost (REAL)
- price_inr (REAL) -- Selling Price
- stock_quantity (INTEGER)
- supplier_id (INTEGER) -- References suppliers
- created_at (TIMESTAMP)

4. tech_packs
- id (INTEGER PRIMARY KEY)
- style_number (TEXT) -- References finished_goods
- fabric_details (TEXT)
- construction (TEXT)
- wash_instructions (TEXT)
- specification_details (TEXT)
- image_url (TEXT)
- image_embedding (vector)
- created_at (TIMESTAMP)

5. sales_orders
- order_number (TEXT PRIMARY KEY)
- buyer_id (INTEGER) -- References buyers
- style_number (TEXT) -- References finished_goods
- quantity (INTEGER)
- order_date (DATE)
- shipment_date (DATE)
- status (TEXT) -- 'Pending', 'Shipped', 'Delivered', 'Cancelled'
- created_at (TIMESTAMP)

6. sales_invoices
- invoice_number (TEXT PRIMARY KEY)
- order_number (TEXT) -- References sales_orders
- amount_inr (REAL)
- currency (TEXT)
- payment_status (TEXT) -- 'Paid', 'Pending', 'Overdue'
- issue_date (DATE)
- due_date (DATE)
- created_at (TIMESTAMP)

Rules:
- Do not query the "image_embedding" column unless explicitly requested.
- Limit output to raw executable SQL query string ONLY. Do not use Markdown styling.
`;

// Heuristic Fallback Rules when OpenAI is not configured
function ruleBasedSqlFallback(question) {
  const q = (question || '').toLowerCase();

  if (q.includes('revenue') || q.includes('sales amount') || q.includes('total amount')) {
    return 'SELECT SUM(amount_inr) AS total_revenue FROM sales_invoices;';
  }
  if (q.includes('top product') || q.includes('finished goods') || q.includes('most stock') || q.includes('stock')) {
    return 'SELECT style_number, category, color, fabric, price_inr, stock_quantity FROM finished_goods ORDER BY stock_quantity DESC LIMIT 5;';
  }
  if (q.includes('supplier')) {
    return 'SELECT name, contact_email, address FROM suppliers LIMIT 5;';
  }
  if (q.includes('buyer') || q.includes('customer')) {
    return 'SELECT name, contact_email, address FROM buyers LIMIT 5;';
  }
  if (q.includes('invoice')) {
    return 'SELECT invoice_number, amount_inr, payment_status, issue_date FROM sales_invoices ORDER BY issue_date DESC LIMIT 5;';
  }
  if (q.includes('order')) {
    return 'SELECT order_number, status, quantity, order_date FROM sales_orders ORDER BY order_date DESC LIMIT 5;';
  }
  if (q.includes('designer') || q.includes('tech pack') || q.includes('specs')) {
    return 'SELECT id, style_number, created_at FROM tech_packs LIMIT 5;';
  }

  // Generic fallback: fetch some finished goods
  return 'SELECT style_number, category, color, fabric, price_inr, stock_quantity FROM finished_goods LIMIT 5;';
}

export async function generateSql(question) {
  if (!openaiClient) {
    console.warn("[WARN] OpenAI API key not configured, using heuristic fallback.");
    return ruleBasedSqlFallback(question);
  }

  try {
    const response = await openaiClient.chat.completions.create({
      model: 'openrouter/free',
      messages: [
        { role: 'system', content: SCHEMA_PROMPT },
        { role: 'user', content: question }
      ],
      temperature: 0.1
    });

    let sql = response.choices[0].message.content.trim();
    if (sql.startsWith('```')) {
      sql = sql.replace(/^```sql?/i, '').replace(/```$/, '').trim();
    }
    return sql;
  } catch (err) {
    console.error("[ERROR] OpenAI text-to-SQL failed:", err.message);
    return ruleBasedSqlFallback(question);
  }
}

export async function generateSummary(question, sql, rows) {
  if (openaiClient) {
    try {
      const response = await openaiClient.chat.completions.create({
        model: 'openrouter/free',
        messages: [
          { role: 'system', content: 'You are a helpful business intelligence assistant. Summarize the returned rows for the query in 1-2 business-friendly sentences.' },
          { 
            role: 'user', 
            content: `User Question: "${question}"\nSQL Query Run: "${sql}"\nResult Rows:\n${JSON.stringify(rows, null, 2)}\nProvide a brief business summary of these results.` 
          }
        ],
        temperature: 0.5
      });
      return response.choices[0].message.content.trim();
    } catch (err) {
      return `Executed SQL successfully. Returned ${rows.length} rows.`;
    }
  } else {
    // Generate simple static summary
    if (rows.length === 0) {
      return "No data found matching your query details.";
    }
    if (rows.length === 1 && Object.keys(rows[0]).length === 1) {
      const colName = Object.keys(rows[0])[0];
      const val = rows[0][colName];
      return `The result for your query shows a value of ${val} for ${colName.replace(/_/g, ' ')}.`;
    }
    return `Query executed successfully and returned ${rows.length} records matching your request.`;
  }
}

File: backend/app/ai_helper.py
import re
from app.config import settings

# Global Vanna client holder
vanna_client = None

def init_vanna():
    """
    Initializes Vanna AI client if keys are available in the configuration.
    """
    global vanna_client
    if settings.VANNA_API_KEY and settings.OPENAI_API_KEY:
        try:
            # Note: We import inside to avoid hard crashes if vanna is not fully installed
            from vanna.openai.openai_chat import OpenAI_Chat
            from vanna.chromadb.chromadb_vector import ChromaDB_VectorStore
            
            class MyVanna(ChromaDB_VectorStore, OpenAI_Chat):
                def __init__(self, config=None):
                    ChromaDB_VectorStore.__init__(self, config=config)
                    OpenAI_Chat.__init__(self, config=config)
                    
            vanna_client = MyVanna(config={
                'api_key': settings.OPENAI_API_KEY,
                'model': 'gpt-4o',
                'path': './vanna_chroma'
            })
            
            # Train on schema if not already trained
            train_vanna(vanna_client)
            print("Successfully initialized Vanna AI with OpenAI GPT-4o.")
        except Exception as e:
            print(f"Failed to initialize Vanna client: {e}. Falling back to Rule-Based SQL Generator.")
            vanna_client = None

def train_vanna(client):
    """
    Trains Vanna client on the ERP schema and sample questions.
    """
    try:
        from app.database import SCHEMA_SQL_PATH
        import os
        if os.path.exists(SCHEMA_SQL_PATH):
            with open(SCHEMA_SQL_PATH, 'r') as f:
                schema_ddl = f.read()
            # Train Vanna on DDL
            client.train(ddl=schema_ddl)
            
            # Train on typical query-SQL pairs (few-shot training)
            client.train(
                question="Show pending invoices above ₹1,000",
                sql="SELECT invoice_number, order_number, amount_inr, payment_status, issue_date FROM sales_invoices WHERE payment_status = 'Pending' AND amount_inr > 1000;"
            )
            client.train(
                question="What is the total revenue by buyer?",
                sql="SELECT b.name AS buyer_name, SUM(i.amount_inr) AS total_revenue FROM buyers b JOIN sales_orders o ON b.id = o.buyer_id JOIN sales_invoices i ON o.order_number = i.order_number GROUP BY b.name ORDER BY total_revenue DESC;"
            )
            client.train(
                question="List style numbers with fabric Silk and GSM above 150",
                sql="SELECT style_number, category, fabric, gsm, price_inr FROM finished_goods WHERE fabric = 'Silk' AND gsm > 150;"
            )
            print("Vanna training complete.")
    except Exception as e:
        print(f"Error training Vanna: {e}")

def generate_sql_query(user_question: str) -> str:
    """
    Generates SQL based on natural language query.
    Uses Vanna AI if configured, otherwise falls back to a rule-based regex mapper.
    """
    global vanna_client
    if vanna_client is None:
        init_vanna()
        
    if vanna_client:
        try:
            return vanna_client.generate_sql(user_question)
        except Exception as e:
            print(f"Vanna generation failed: {e}. Using rule-based fallback.")
            
    # Rule-based fallback (regex matching for standard ERP questions)
    q = user_question.lower().strip()
    
    # 1. "Show pending invoices above ₹1,000"
    if "pending" in q and "invoice" in q and ("above" in q or "greater" in q or ">" in q):
        # Extract number
        nums = re.findall(r'\d+', q.replace(',', ''))
        amount = nums[0] if nums else "1000"
        return f"SELECT invoice_number, order_number, amount_inr, payment_status, due_date FROM sales_invoices WHERE payment_status = 'Pending' AND amount_inr > {amount} ORDER BY amount_inr DESC;"
        
    # 2. "What is the total revenue by buyer?" or "revenue by buyer"
    if "revenue" in q and "buyer" in q:
        return """SELECT b.name AS buyer_name, 
       COUNT(DISTINCT o.order_number) AS total_orders,
       SUM(i.amount_inr) AS total_revenue_inr
FROM buyers b
JOIN sales_orders o ON b.id = o.buyer_id
JOIN sales_invoices i ON o.order_number = i.order_number
WHERE i.payment_status = 'Paid'
GROUP BY b.name
ORDER BY total_revenue_inr DESC;"""

    # 3. "List style numbers with fabric Silk and GSM above 150"
    if "style" in q and "fabric" in q and "gsm" in q:
        # Extract fabric (e.g. Silk, Cotton)
        fabrics = ['cotton', 'silk', 'linen', 'polyester', 'wool', 'denim', 'rayon', 'nylon']
        found_fabric = "Silk"
        for f in fabrics:
            if f in q:
                found_fabric = f.capitalize()
                break
        # Extract GSM limit
        nums = re.findall(r'\d+', q)
        gsm_val = nums[0] if nums else "150"
        return f"SELECT style_number, category, fabric, gsm, color, price_inr, stock_quantity FROM finished_goods WHERE LOWER(fabric) = '{found_fabric.lower()}' AND gsm > {gsm_val} ORDER BY gsm ASC;"

    # 4. "Show all suppliers" / "List suppliers"
    if "supplier" in q:
        return "SELECT id, name, contact_email, phone, address FROM suppliers ORDER BY name ASC;"
        
    # 5. "Show all buyers" / "List buyers"
    if "buyer" in q:
        return "SELECT id, name, contact_email, phone, address FROM buyers ORDER BY name ASC;"

    # 6. "Total sales" / "Total revenue"
    if "total revenue" in q or "total sales" in q:
        return "SELECT SUM(amount_inr) AS total_revenue_inr, COUNT(invoice_number) AS total_invoices FROM sales_invoices WHERE payment_status = 'Paid';"

    # 7. "How many orders are pending?" or "pending orders"
    if "pending" in q and "order" in q:
        return "SELECT order_number, buyer_id, style_number, quantity, order_date, status FROM sales_orders WHERE status = 'Pending' ORDER BY order_date DESC;"

    # Default fallback - return a query showing recently created finished goods style numbers
    return "SELECT style_number, category, fabric, gsm, color, price_inr, stock_quantity FROM finished_goods LIMIT 10;"

def generate_text_answer(question: str, sql: str, results: list) -> str:
    """
    Generates a natural language summary of the SQL query results.
    """
    if not results:
        return "No records were found matching your criteria."
        
    q = question.lower()
    
    # 1. Invoices answer
    if "pending" in q and "invoice" in q:
        total_amt = sum(row.get('amount_inr', 0) for row in results)
        count = len(results)
        return f"Found **{count} pending invoices** totaling **₹{total_amt:,.2f}**. The highest pending invoice is **{results[0].get('invoice_number')}** for **₹{results[0].get('amount_inr', 0):,.2f}**."
        
    # 2. Revenue answer
    if "revenue" in q and "buyer" in q:
        top_buyer = results[0].get('buyer_name', 'N/A')
        top_rev = results[0].get('total_revenue_inr', 0)
        return f"The buyer generating the highest revenue is **{top_buyer}** with **₹{top_rev:,.2f}** in paid sales. A total of **{len(results)} buyers** have recordable purchases."
        
    # 3. Fabrics/GSM answer
    if "fabric" in q or "gsm" in q:
        return f"Found **{len(results)} styles** matching your fabric/GSM request. The matching style numbers are: {', '.join([row.get('style_number') for row in results[:5]])}{' and others' if len(results) > 5 else ''}."
        
    # Default text response summarizing fields
    return f"Query returned **{len(results)} rows**. Columns returned: {', '.join(results[0].keys())}."
